sdk/python/feast/infra/utils/aws_utils.py

```diff
@@ -0,0 +1,286 @@
+    Raises RedshiftCredentialsError if the statement couldn't be executed due to the validation error.
+
+    :param redshift_data_client: Redshift Data API Service client
+    :param cluster_id: Redshift Cluster Identifier
+    :param database: Redshift Database Name
+    :param user: Redshift username
+    :param query: The SQL query to execute
+    :return: JSON response
+    """
+    try:
+        return redshift_data_client.execute_statement(
+            ClusterIdentifier=cluster_id, Database=database, DbUser=user, Sql=query,
+        )
+    except ClientError as e:
+        if e.response["Error"]["Code"] == "ValidationException":
+            raise RedshiftCredentialsError() from e
+        raise
+
+
+class RedshiftStatementNotFinishedError(Exception):
+    pass
+
+
+@retry(
+    wait=wait_exponential(multiplier=0.1, max=30),
+    retry=retry_if_exception_type(RedshiftStatementNotFinishedError),
+)
+def wait_for_redshift_statement(redshift_data_client, statement: dict) -> None:
+    """
+    Waits for the Redshift statement to finish. Raises RedshiftQueryError if the statement didn't succeed.
+
+    We use exponential backoff for checking the query state until it's not running. The backoff starts with
+    0.1 seconds and doubles exponentially until reaching 30 seconds, at which point the backoff is fixed.
+
+    :param redshift_data_client: Redshift Data API Service client
+    :param statement: The redshift statement to wait for (result of execute_redshift_statement)
+    """
+    desc = redshift_data_client.describe_statement(Id=statement["Id"])
+    if desc["Status"] in ("SUBMITTED", "STARTED", "PICKED"):
+        raise RedshiftStatementNotFinishedError  # Retry
+    if desc["Status"] != "FINISHED":
+        raise RedshiftQueryError(desc)  # Don't retry. Raise exception.
+
+
+def execute_redshift_statement(
+    redshift_data_client, cluster_id: str, database: str, user: str, query: str
+) -> str:
+    """
+    Execute Redshift statement synchronously. Waits for the query to finish.
+    Raises RedshiftCredentialsError if the statement couldn't be executed due to the validation error.
+    Raises RedshiftQueryError if the query runs but finishes with errors.
+
+    :param redshift_data_client: Redshift Data API Service client
+    :param cluster_id: Redshift Cluster Identifier
+    :param database: Redshift Database Name
+    :param user: Redshift username
+    :param query: The SQL query to execute
+    :return: Statement ID
+    """
+    statement = execute_redshift_statement_async(
+        redshift_data_client, cluster_id, database, user, query
+    )
+    wait_for_redshift_statement(redshift_data_client, statement)
+    return statement["Id"]
+
+
+def get_redshift_statement_result(redshift_data_client, statement_id: str) -> dict:
+    """ Get the Redshift statement result """
+    return redshift_data_client.get_statement_result(Id=statement_id)
+
+
+def copy_df_to_redshift(
```

Comment:
I think it's a bit confusing to call this method `copy`. Should we use `upload_` to be consistent with BigQuery?